scanner/matcher/updater/vuln/updater.go

```diff
@@ -326,17 +335,127 @@ func (u *Updater) runUpdate(ctx context.Context) error {
 		return err
 	}
 
-	if err := u.metadataStore.SetLastVulnerabilityUpdate(ctx, timestamp); err != nil {
+	if err := u.metadataStore.SetLastVulnerabilityUpdate(ctx, timestamp, updateName); err != nil {
 		return err
 	}
 
-	if u.initialized.CompareAndSwap(false, true) {
-		zlog.Info(ctx).Msg("finished initial updater run: setting updater to initialized")
-	}
+	return nil
+}
+
+// runSingleBundleUpdate updates the vulnerability data.
+func (u *Updater) runMultiBundleUpdate(ctx context.Context) error {
+	ctx = zlog.ContextWithValues(ctx, "component", "matcher/updater/vuln/Updater.runMultiBundleUpdate")
 
+	updateTime, err := u.metadataStore.GetLastVulnerabilityUpdate(ctx)
+	if err != nil {
+		zlog.Debug(ctx).
+			Err(err).
+			Msg("did not get previous vuln update timestamp")
+		return err
+	}
 	zlog.Info(ctx).
-		Str("timestamp", timestamp.Format(http.TimeFormat)).
-		Msg("new vuln update")
+		Time("timestamp", updateTime).
+		Msg("previous vuln update")
+
+	zipFile, zipTime, err := u.fetch(ctx, updateTime)
+	if err != nil {
+		return err
+	}
+	if zipFile == nil {
+		// Nothing to update at this time.
+		zlog.Info(ctx).Msg("no new vulnerability update")
+		return nil
+	}
+	defer func() {
+		if err := zipFile.Close(); err != nil {
+			zlog.Error(ctx).Err(err).Msg("closing temp update file")
+		}
+		if err := os.Remove(zipFile.Name()); err != nil {
+			zlog.Error(ctx).Err(err).Msgf("removing temp update file %q", zipFile.Name())
+		}
+	}()
+
+	// Open the .zip file
+	zipInfo, err := zipFile.Stat()
+	if err != nil {
+		return err
+	}
+	zipReader, err := zip.NewReader(zipFile, zipInfo.Size())
+	if err != nil {
+		return err
+	}
+
+	// Iterate through each vulnerability bundle in the .zip archive
+	names := make([]string, 0, len(zipReader.File))
+	for i := range zipReader.File {
+		bundleF := zipReader.File[i]
+		names = append(names, bundleF.Name)
+		ctx := zlog.ContextWithValues(ctx, "bundle", bundleF.Name)
+		zlog.Info(ctx).Msg("starting bundle update")
+		if err := u.updateBundle(ctx, bundleF, zipTime); err != nil {
+			zlog.Error(ctx).Err(err).Msg("updating bundle failed")
+			return fmt.Errorf("updating bundle %s: %w", bundleF.Name, err)
+		}
+		zlog.Info(ctx).Msg("completed bundle update")
```

Comment:
Should this line be outside of the loop with message: _completed all bundles update_, means all bundles have been updated? Because if one bundle fails, the current update process will quit with error until next round of update starts.

Reply:
> Because if one bundle fails, the current update process will quit with error until next round of update starts.

I don't see how quitting the whole update process in the middle would make this log message inappropriate. I may be confused because I see two log entries: One for when the bundle update fails and another for when it succeeds. And finally, there are log messages for when the whole update process starts and ends. Can you help me understand what is your concern?

Reply:
My point is: it might be better to have `zlog.Info(ctx).Msg("starting bundle update")` in line 390 and `zlog.Info(ctx).Msg("completed bundle update")` in line 401. We can reduce noise by logging only when all bundles are successfully imported. Since an error is returned if any bundle in the archive fails to import, there is no need to log successful imports until all bundles are completely processed.